pkg/cloud/services/compute/instance.go

```diff
@@ -46,57 +46,150 @@ const (
+						return subnet.NetworkID, nil
+					}
 				}
-				if len(netIDs) > 1 {
-					return nil, fmt.Errorf("network filter for port %s returns more than one result", port.NameSuffix)
-				} else if len(netIDs) == 0 {
-					return nil, fmt.Errorf("network filter for port %s returns no networks", port.NameSuffix)
-				}
-				netID = netIDs[0]
+
+				// TODO: This should set the machine to failed
+				return "", fmt.Errorf("port %d has no network and unable to infer from fixed IPs", portIdx)
+			}()
+			if err != nil {
+				return nil, err
 			}
-			nets = append(nets, infrav1.Network{
-				ID:       netID,
-				Subnet:   &infrav1.Subnet{},
-				PortOpts: port,
-			})
-		} else {
-			nets = append(nets, infrav1.Network{
-				ID: openStackCluster.Status.Network.ID,
-				Subnet: &infrav1.Subnet{
-					ID: openStackCluster.Status.Network.Subnet.ID,
-				},
-				PortOpts: port,
-			})
+
+			port.Network = &infrav1.NetworkFilter{
+				ID: networkID,
+			}
+
+		// Network is defined by ID
+		case port.Network.ID == "":
+			networkingService, err := s.getNetworkingService()
+			if err != nil {
+				return nil, err
+			}
+
+			netIDs, err := networkingService.GetNetworkIDsByFilter(port.Network.ToListOpt())
+			if err != nil {
+				return nil, err
+			}
+
+			// TODO: These errors should set the machine to failed
+			if len(netIDs) > 1 {
+				return nil, fmt.Errorf("network filter for port %d returns more than one result", portIdx)
+			} else if len(netIDs) == 0 {
+				return nil, fmt.Errorf("network filter for port %d returns no networks", portIdx)
+			}
+
+			port.Network.ID = netIDs[0]
+
+		// Network is already defined by ID
+		default:
+			// Nothing to do
+		}
+
+		normalizedPorts = append(normalizedPorts, port)
+	}
+	return normalizedPorts, nil
+}
+
+// constructNetworks builds an array of networks from the network, subnet and ports items in the instance spec.
+// If no networks or ports are in the spec, returns a single network item for a network connection to the default cluster network.
+func (s *Service) constructNetworks(openStackCluster *infrav1.OpenStackCluster, instanceSpec *InstanceSpec) ([]infrav1.Network, error) {
+	nets, err := s.getServerNetworks(instanceSpec.Networks)
+	if err != nil {
+		return nil, err
+	}
+
+	// Ensure user-specified ports have all required fields
+	ports, err := s.normalizePorts(instanceSpec.Ports, openStackCluster, instanceSpec)
+	if err != nil {
+		return nil, err
+	}
+	for i := range ports {
+		port := ports[i]
+		nets = append(nets, infrav1.Network{
+			ID:       port.Network.ID,
+			Subnet:   &infrav1.Subnet{},
+			PortOpts: &port,
+		})
+	}
+
+	// trunk support is required if the instance or any port has trunk enabled
+	portUsesTrunk := func() bool {
+		for _, port := range instanceSpec.Ports {
+			if port.Trunk != nil && *port.Trunk {
+				return true
+			}
+		}
+		return false
+	}
+	if instanceSpec.Trunk || portUsesTrunk() {
```

Comment:
If you have `instanceSpec.Trunk=true`, but all the ports override that with `false`, then I guess we don't need trunk extension. We could only check `portUsesTrunk()` here.

Reply:
Nice edge case! This is actually here because we create the default port down below based on instanceSpec.Trunk, so neither is quite correct. I'll have a think.